lib/action_client/middleware/parser.rb

```diff
@@ -0,0 +1,61 @@
+module ActionClient
+  module Middleware
+    class Parser
+      JsonParser = proc do |body|
+        JSON.parse(body, object_class: HashWithIndifferentAccess)
+      end
+      XmlParser = proc do |body|
+        Nokogiri::XML.parse(body).tap do |document|
+          document.validate
+
+          document.errors.each do |error|
+            if error.is_a?(Nokogiri::XML::SyntaxError)
+              raise error
+            end
+          end
+        end
+      end
+      NullParser = -> (body) { body }
+
+      class_attribute :parsers, default: {
+        "application/json": JsonParser,
+        "application/xml": XmlParser,
+      }.with_indifferent_access
+
+      def initialize(app, configuration = {})
+        @app = app
+        @parsers = self.class.parsers.merge(configuration.fetch(:parsers, {}))
+      end
+
+      def call(env)
+        status, headers, body_proxy = @app.call(env)
+        body = body_proxy.each(&:yield_self).join
+        content_type = headers[Rack::CONTENT_TYPE].to_s
+
+        if body.present?
+          parser = fetch_parser_for_content_type(content_type)
+
+          begin
+            [ status, headers, parser.call(body) ]
+          rescue StandardError => error
+            raise ActionClient::ParseError.new(error, body, content_type)
+          end
+        else
+          [ status, headers, body ]
+        end
+      end
+
+      private
+
+      def fetch_parser_for_content_type(content_type)
+        _, content_parser = @parsers.detect do |key, parser|
+          if content_type.to_s.starts_with?(key.to_s)
```

Comment:
Using `#starts_with?` here means that we implicitly support parsers for types without a subtype (e.g. `text` or `image`).

Is that something we should document and add test coverage for?

Reply:
I've pushed up tests and documentation.

Is this quirk a positive feature, or a liability? 

If the problem is general, re-usable parsers, we could potentially support regular expressions as well.

Reply:
You could be right about it being a liability. Looking at [Mozilla's list of common MIME types][1] there are some edge cases where one MIME type starts with another: `application/x-sh` for Bourne shell scripts probably shouldn't match a parser for `application/x-shockwave-flash`. Perhaps more likely, there are several types where different versions of a file format like `application/x-bzip` and `application/x-bzip2`.

We could use Rails' `Mime::Type` class to ensure we are only getting fully qualified types. This would also have the added benefit of giving us support for synonyms (e.g. `text/xml`, `application/xml`, `application/x-xml`).

At the point where we set the config, we'd need something like:

```ruby
@parsers = self.class.parsers.merge(
  configuration.fetch(:parsers, {})
    .transform_keys { |type| Mime::Type.lookup(type).to_s }
)
```

As a result, this method gets much shorter:

```ruby
def fetch_parser_for_content_type(content_type)
  @parsers.fetch(Mime::Type.lookup(content_type).to_s) { :itself.to_proc }
end
```

[1]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types